pkg/bgpv1/agent/controller.go

```diff
@@ -103,80 +104,63 @@ func NewController(params ControllerParams) (*Controller, error) {
 		BGPMgr:             params.RouteMgr,
 		PolicyResource:     params.PolicyResource,
 		CiliumNodeResource: params.LocalCiliumNodeResource,
 	}
 
-	params.Lifecycle.Append(&c)
-
-	return &c, nil
-}
+	jobGroup := params.JobRegistry.NewGroup(
+		params.Scope,
+		job.WithLogger(log),
+		job.WithPprofLabels(pprof.Labels("cell", "bgp-cp")),
+	)
 
-// Start is called by hive after all of our dependencies have been started.
-func (c *Controller) Start(startCtx hive.HookContext) error {
-	policyStore, err := c.PolicyResource.Store(startCtx)
-	if err != nil {
-		return fmt.Errorf("PolicyResource.Store(): %w", err)
-	}
-	c.PolicyLister = policyListerFunc(func() ([]*v2alpha1api.CiliumBGPPeeringPolicy, error) {
-		return policyStore.List(), nil
-	})
-
-	c.workerpool = workerpool.New(3)
-
-	c.workerpool.Submit("policy-observer", func(ctx context.Context) error {
-		for ev := range c.PolicyResource.Events(ctx) {
-			switch ev.Kind {
-			case resource.Upsert, resource.Delete:
-				// Signal the reconciliation logic.
-				c.Sig.Event(struct{}{})
+	jobGroup.Add(
+		job.OneShot("bgp-policy-observer", func(ctx context.Context, health cell.HealthReporter) (err error) {
+			for ev := range c.PolicyResource.Events(ctx) {
+				switch ev.Kind {
+				case resource.Upsert, resource.Delete:
+					// Signal the reconciliation logic.
+					c.Sig.Event(struct{}{})
+				}
+				ev.Done(nil)
 			}
-			ev.Done(nil)
-		}
-		return nil
-	})
-
-	c.workerpool.Submit("cilium-node-observer", func(ctx context.Context) error {
-		for ev := range c.CiliumNodeResource.Events(ctx) {
-			switch ev.Kind {
-			case resource.Upsert:
-				// Set the local CiliumNode.
-				c.LocalCiliumNode = ev.Object
-				// Signal the reconciliation logic.
-				c.Sig.Event(struct{}{})
+			return nil
+		}),
+
+		job.OneShot("cilium-node-observer", func(ctx context.Context, health cell.HealthReporter) (err error) {
+			for ev := range c.CiliumNodeResource.Events(ctx) {
+				switch ev.Kind {
+				case resource.Upsert:
+					// Set the local CiliumNode.
+					c.LocalCiliumNode = ev.Object
+					// Signal the reconciliation logic.
+					c.Sig.Event(struct{}{})
+				}
+				ev.Done(nil)
 			}
-			ev.Done(nil)
-		}
-		return nil
-	})
+			return nil
+		}),
 
-	c.workerpool.Submit("controller", func(ctx context.Context) error {
-		c.Run(ctx)
-		return nil
-	})
-
-	return nil
-}
-
-// Stop is called by hive upon shutdown, after all of our dependants have been stopped.
-// We should perform a graceful shutdown and return as soon as done or when the stop context is done.
-func (c *Controller) Stop(ctx hive.HookContext) error {
-	doneChan := make(chan struct{})
-	go func() {
-		c.workerpool.Close()
-		close(doneChan)
-	}()
-
-	select {
-	case <-ctx.Done():
-		return ctx.Err()
-	case <-doneChan:
-	}
+		job.OneShot("bgp-controller", func(ctx context.Context, health cell.HealthReporter) (err error) {
+			// initialize PolicyLister used in the controller
+			policyStore, err := c.PolicyResource.Store(ctx)
```

Comment:
Maybe store initialization can be moved into separate method and on error we can set health report to degraded or failed.

Reply:
> since such failures are not going to stop agent from starting up

`job.WithShutdown()` that I used in the jobs that can potentially fail actually causes the agent to stop if the job is not successful in the specified num. of attempts. I think it makes sense to stop on failure, as the BGP CP will not function properly without stores being initialized, and it also was the existing behavior before this change.

If we decided to run BGP CP even without stores being initialized, we would need to retry forever to be able to get into a consistent state eventually.

BTW, if we did not shutdown, the hive infra would automatically set the health of a failed job to Degraded:
https://github.com/cilium/cilium/blob/12884735a6605365d3f7ba6067bfd2976f5bfaa6/pkg/hive/job/job.go#L357-L361

Reply:
How long is the retry now ? Should we wait forever instead ( or large enough retry interval ) ? If operator does not install CRD in time ( in case of overloaded cluster ), BGP CP will retry few times and then fail - in which case we are back to same situation. 

I guess it's usual question of how long should we wait before declaring panic.

Reply:
In case of missing CRDs we actually already wait forever, as resource.Store blocks until CRDs are installed and caches are synced (or provided context is cancelled).

The retry that I put in for these jobs is more about retrying some other potential errors (I guess quite hypothetical, not sure what those can be). Those retries would be quite quick in that case `DefaultControllerRateLimiter` starts exponential backoff at `5ms`.

Reply:
Ah i see, make sense in that case. Thanks for digging into this.